src/derivative_operators/convolutions.jl

```diff
@@ -59,18 +95,26 @@ end
 
 # Against A BC-padded vector, specialize the computation to explicitly use the left, right, and middle parts
 function convolve_interior!(x_temp::AbstractVector{T}, _x::BoundaryPaddedVector, A::DerivativeOperator) where {T<:Real}
+    @assert length(x_temp) == length(_x.u)
     stencil = A.stencil_coefs
     coeff   = A.coefficients
-    x = _x.u
-    mid = div(A.stencil_length,2) + 1
+    # Upwind operators have a non-centred stencil
+    if use_winding(A)
+        mid = 1 + A.stencil_length%2
+    else
+        mid = div(A.stencil_length,2)
+    end
+
     # Just do the middle parts
-    for i in (2+A.boundary_point_count) : (length(x_temp)-A.boundary_point_count)-1
```

Comment:
We needed to start indexing from 2 here since one interior stencil the left will always use the ghost point _x.l. There is an analogous situation for the rightmost interior stencil with _x.r. Consequently, we moved the special handling of _x.l and _x.r to the boundary convolutions.

Reply:
If boundary_point_count is 0, then the loop in `convolve_left_BC!` will not run only. So the `interior_convolve!` function has to handle the boundary also.